### DISCUSSION TITLE
The quickjs.h can't be included in a c++ compilation unit, can it?

### DISCUSSION THREAD

**Discussion by Please-just-dont** [quickjs-ng/quickjs#538] [Q&A]
It looks like this Javascript interpreter was written in a style of C that isn't compatible with C++ and thus won't compile.  JS_MKVAL macro is underlined:

static js_force_inline JSValue JS_NewBool(JSContext *ctx, JS_BOOL val)
{
    return JS_MKVAL(JS_TAG_BOOL, (val != 0));
}

Error	C4576	a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax	

I looked at the macro, I'm pretty sure I ran into this problem before and I'm pretty sure it has something to do with a C feature, can't remember what it's called anymore, and I couldn't find it from a Google search, but it looks like this (MyType) { .member1 = 2, .member2 = 3 } . That gnarly looking cast thingy before the initialisation, it's a newer C feature that won't compile in C++.

**Comment by ericmelse** [comment]
Hi Please-just-dont,

I have put your question text to the [ Code: Java, C/C++ } GPT engine. This the answer that I got:

==============

The issue you're encountering stems from the fact that QuickJS uses **designated initializers** and **compound literals**, which are features available in C99 but not in C++. In C99, you can initialize structures using a syntax like this:

`(MyType){.member1 = 2, .member2 = 3};`

However, C++ does not support this syntax in the same way, which is why you're running into the error:
`a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax`.

**Solution**
To make this compatible with C++, you'll need to modify the code to avoid using C99's designated initializers and compound literals.
Let's start by analyzing the macro JS_MKVAL.

**What the Macro Looks Like in C**
It's likely that the macro JS_MKVAL is defined something like this:

`#define JS_MKVAL(tag, val) ((JSValue){.tag = (tag), .val = (val)})`

This syntax is allowed in C but not in C++.
To resolve this, you'll need to rewrite it in a way that's valid for both C and C++.

**Fix: Convert C-style Compound Literals to C++-style Initialization**
In C++, you can't directly use the designated initializer syntax.
Instead, you'll need to use a constructor-style initialization. 
Here's what you can do:

1. If `JSValue` is a **struct** or **class**, you'll need to provide a constructor that accepts both `tag` and `val`.
2. Modify the macro so that it uses C++-compliant syntax.

**Steps to Fix**:

1. **Modify** `JSValue` **struct/class**:
Instead of relying on designated initializers, create a constructor in `JSValue` that takes the `tag` and `val` as parameters:
```
struct JSValue {
    int tag;
    int val;

    // Constructor to handle the initialization
    JSValue(int t, int v) : tag(t), val(v) {}
};
```
2. **Update the Macro**:
Change the `JS_MKVAL` macro to use the constructor-style syntax, which is valid in both C++ and C:
`#define JS_MKVAL(tag, val) (JSValue(tag, val))`

3. **Update the Function**:
Your function `JS_NewBool` should now look like this:
```
static js_force_inline JSValue JS_NewBool(JSContext *ctx, JS_BOOL val)
{
    return JS_MKVAL(JS_TAG_BOOL, (val != 0));
}
```
This avoids the C99-specific syntax that causes the issue, and will compile in both C and C++.

**Summary of Key Changes**
- **Struct Constructor**: Added a constructor to `JSValue` for tag and val.
- **Macro Update**: Changed the `JS_MKVAL` macro to use constructor-style initialization.
This should resolve the error you're encountering in C++ while keeping the code functional.

**Additional Suggestions**:
If `JSValue` is more complex or has multiple fields, you may need to ensure that any initialization logic or further functions are adapted for compatibility across both C and C++.

==============

I hope that this is helpful to follow up on this issue and solve it.

Best,
Eric

**Reply by saghul** [reply]
What commit did you test?

**Reply by ericmelse** [reply]
Dear Saghul, personally I did not test anything by my friend programmer did.
His comments are:

Yes, indeed, I encountered a problem compiling c99
So, I decided that at the moment it would be easier to build QuickJS as a separate application and test whether it is suitable for our task, and if it is suitable, then try to build it as a separate static lib.

My suggestion is to look into solving this problem as suggested by the ChatGPT coding bot.

**Reply by saghul** [reply]
PRs would be welcome, but we already landed some to that end, so I think it should already work.

**Comment by Please-just-dont** [comment]
@saghul Hi, this is pretty much the latest version I tested, it was from like 7 days ago, doesn't compile in C++ for the reason given, it's using compound literals in the header and therefore can't be included into c++

**Reply by saghul** [reply]
Got it. Care to send a PR?

**Reply by Please-just-dont** [reply]
I'm not familiar with this project and I wouldn't feel comfortable changing the code on it.
